pandas/core/frame.py

```diff
@@ -5355,6 +5355,18 @@ def reindex(
         limit: int | None = None,
         tolerance=None,
     ) -> DataFrame:
+        # Automatically detect matching level when reindexing from Index to MultiIndex.
+        # This prevents values from being incorrectly set to NaN when the source index
+        # name matches a level name in the target MultiIndex. Only applies when source
+        # is not already a MultiIndex.
+        if (
+            level is None
+            and index is not None
+            and isinstance(index, MultiIndex)
+            and not isinstance(self.index, MultiIndex)
+            and self.index.name in index.names
+        ):
+            level = self.index.name
```

Comment:
Is it possible to combine this logic with the `series.py` logic and push it into `generic.py`?